### ISSUE TITLE
Offer Singularity / Apptainer container

### ISSUE THREAD

**Issue by nikosavola** [awslabs/palace#60]
I suggest supporting Singularity (or nowadays also Apptainer) containers in addition to the Spack manager currently supported. At least I find Singularity more widely supported than Spack.

It seems [easy to containerise Spack projects](https://spack.readthedocs.io/en/latest/containers.html) such that a user can easily do it themselves, but would be nice a purpose-built definition file would be provided or built in CI.

**Comment by sebastiangrimberg** [comment]
Hi @nikosavola, thanks for your suggestion. Indeed it does not seems too difficult to generate container recipes for Docker or Singularity based on a package which supports Spack. We do not use containers often in our workflows and prefer to build and run Palace without containerization, but I understand this can be of interest to some users. I would welcome a PR to add this functionality to the repository if you had something in mind. Probably just supplying the container recipe and not actually building and registering the container would be sufficient? I know projects like MFEM and libCEED both have some degree of automated containerization support, perhaps they would be a good place to start for this.

**Comment by nikosavola** [comment]
> Hi @nikosavola, thanks for your suggestion. Indeed it does not seems too difficult to generate container recipes for Docker or Singularity based on a package which supports Spack. We do not use containers often in our workflows and prefer to build and run Palace without containerization, but I understand this can be of interest to some users. I would welcome a PR to add this functionality to the repository if you had something in mind. Probably just supplying the container recipe and not actually building and registering the container would be sufficient? I know projects like MFEM and libCEED both have some degree of automated containerization support, perhaps they would be a good place to start for this.

Cool, I might make a contribution for a recipe in that case 👍.

**Comment by nikosavola** [comment]
@sebastiangrimberg, I'm having some problems compiling on CentOS7. Namely, compiling PETSc and SLEPc is giving out the following. Do you think something is wrong in the CMakeLists?
```
-- ================== Finished dependency configuration ===================
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.27.1")
-- Checking for module 'PETSc'
--   Found PETSc, version 3.19.99
-- Found PETSc: 3.19.99
-- PETSc test program - Failed
-- PETSc test program with static linkage - Success
-- Checking for module 'SLEPc'
--   No package 'SLEPc' found
-- Checking for module 'slepc'
--   Found slepc, version 3.19.99
-- Found SLEPc: 3.19.99
-- SLEPc test program - Failed
-- SLEPc test program with static linkage - Failed
CMake Error at CMakeLists.txt:116 (message):
  SLEPc could not be found, be sure to set SLEPC_DIR


-- Configuring incomplete, errors occurred!
make[2]: *** [CMakeFiles/palace.dir/build.make:93: palace-cmake/src/palace-stamp/palace-configure] Error 1
make[1]: *** [CMakeFiles/Makefile2:112: CMakeFiles/palace.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

This occurs with the following minimal working example. Compiling mpich takes a while, 
```bash
Bootstrap: docker
From: centos:7.9.2009

%post -c /bin/bash
    #!/bin/bash

    export MARCH="native"

    yum -y upgrade
    yum -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm

    yum -y install centos-release-scl
    yum -y install devtoolset-11
    yum -y install rh-git218
    yum -y install openssl-devel
    yum -y install cmake3
    source /opt/rh/devtoolset-11/enable
    source /opt/rh/rh-git218/enable
    yum -y install libcurl
    
    # Install additional stuff
    yum -y install wget lbzip2 libsndfile numactl zlib perl

    # Install mpich
    wget -qO- http://www.mpich.org/static/downloads/4.0.2/mpich-4.0.2.tar.gz | tar xvz
    cd mpich-4.0.2 || exit
    ./configure --enable-fast=all,O3 --prefix=/usr FFLAGS="-std=legacy" FCFLAGS="-std=legacy"
    make -j
    make install
    ldconfig
    
    # Install newer cmake, needed for Palace dependencies
    wget -qO- https://cmake.org/files/v3.26/cmake-3.26.4.tar.gz | tar xvz
    cd cmake-3.26.4 || exit
    ./bootstrap --prefix=/usr/local -- -DCMAKE_BUILD_TYPE:STRING=Release
    make -j
    make install
    cd ..
    rm -rf cmake-3.26.4

    yum -y install openblas-devel
    yum -y install m4
    yum -y install curl-devel
    yum -y install hdf5-devel

    mkdir /opt/src
    cd /opt/src || exit
   
    # Compile Palace
    git clone https://github.com/awslabs/palace.git --recurse-submodules
    cd palace || exit
    mkdir build
    cd build || exit
    cmake ../ -DCMAKE_INSTALL_PREFIX=/opt/palace \
            -DPALACE_WITH_OPENMP:BOOL=TRUE \
            -DCMAKE_CXX_FLAGS="-O3 -fopenmp -funroll-loops -march=$MARCH" \
            -DCMAKE_C_FLAGS="-O3 -fopenmp -funroll-loops -march=$MARCH -D_POSIX_C_SOURCE=199309L" \
            -DCMAKE_Fortran_FLAGS="-O3 -fPIC -funroll-loops -march=$MARCH" \
            -DPALACE_WITH_MUMPS:BOOL=TRUE \
            -DPALACE_WITH_SUPERLU:BOOL=FALSE \
            -DOPENBLAS_DIR="/usr/include/openblas/"
    make -j
```

**Comment by sebastiangrimberg** [comment]
In order to try to debug this, can you uncomment these lines in https://github.com/awslabs/palace/blob/main/cmake/PkgConfigHelpers.cmake:

https://github.com/awslabs/palace/blob/1631367e57162d4fdd8ebe284e5304a5de14ee74/cmake/PkgConfigHelpers.cmake#L61

https://github.com/awslabs/palace/blob/1631367e57162d4fdd8ebe284e5304a5de14ee74/cmake/PkgConfigHelpers.cmake#L99

Then the error message should be a bit more descriptive and we can go from there to try to resolve it. Thanks!

**Comment by nikosavola** [comment]
I then get the following
```
-- ================== Finished dependency configuration ===================
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.27.1")
-- Checking for module 'PETSc'
--   Found PETSc, version 3.19.99
-- Found PETSc: 3.19.99
-- PETSC_TEST_COMPILE_OUTPUT: Change Dir: /opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/local/bin/cmake -E env VERBOSE=1 /opt/rh/devtoolset-11/root/usr/bin/gmake -f Makefile cmTC_44785/fast && gmake[3]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
/opt/rh/devtoolset-11/root/usr/bin/gmake  -f CMakeFiles/cmTC_44785.dir/build.make CMakeFiles/cmTC_44785.dir/build
gmake[4]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_44785.dir/petsc_test_lib.cpp.o
/opt/rh/devtoolset-11/root/usr/bin/c++  -isystem /opt/palace/include -O3 -fopenmp -funroll-loops -march=native  -std=c++17 -fopenmp -o CMakeFiles/cmTC_44785.dir/petsc_test_lib.cpp.o -c /opt/src/palace/build/palace-build/CMakeFiles/try_run/petsc_test_lib.cpp
/opt/src/palace/build/palace-build/CMakeFiles/try_run/petsc_test_lib.cpp:7:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
    7 |   PetscCall(PetscInitialize(&argc, &argv, PETSC_NULL, PETSC_NULL));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/src/palace/build/palace-build/CMakeFiles/try_run/petsc_test_lib.cpp:7:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
Linking CXX executable cmTC_44785
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_44785.dir/link.txt --verbose=1
/opt/rh/devtoolset-11/root/usr/bin/c++ -O3 -fopenmp -funroll-loops -march=native  -Wl,-rpath -Wl,/usr/lib -Wl,--enable-new-dtags CMakeFiles/cmTC_44785.dir/petsc_test_lib.cpp.o -o cmTC_44785  /opt/palace/lib/libpetsc.a /usr/lib/libmpicxx.so /usr/lib/libmpi.so
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: /opt/palace/lib/libpetsc.a(dlimpl.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: /lib64/libdl.so.2: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
gmake[4]: *** [CMakeFiles/cmTC_44785.dir/build.make:102: cmTC_44785] Error 1
gmake[4]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
gmake[3]: *** [Makefile:127: cmTC_44785/fast] Error 2
gmake[3]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'


-- PETSC_TEST_OUTPUT:
-- PETSc test program - Failed
-- PETSC_TEST_COMPILE_OUTPUT: Change Dir: /opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/local/bin/cmake -E env VERBOSE=1 /opt/rh/devtoolset-11/root/usr/bin/gmake -f Makefile cmTC_abca4/fast && gmake[3]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
/opt/rh/devtoolset-11/root/usr/bin/gmake  -f CMakeFiles/cmTC_abca4.dir/build.make CMakeFiles/cmTC_abca4.dir/build
gmake[4]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_abca4.dir/petsc_test_lib.cpp.o
/opt/rh/devtoolset-11/root/usr/bin/c++  -isystem /opt/palace/include -O3 -fopenmp -funroll-loops -march=native  -std=c++17 -fopenmp -o CMakeFiles/cmTC_abca4.dir/petsc_test_lib.cpp.o -c /opt/src/palace/build/palace-build/CMakeFiles/try_run/petsc_test_lib.cpp
/opt/src/palace/build/palace-build/CMakeFiles/try_run/petsc_test_lib.cpp:7:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
    7 |   PetscCall(PetscInitialize(&argc, &argv, PETSC_NULL, PETSC_NULL));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/src/palace/build/palace-build/CMakeFiles/try_run/petsc_test_lib.cpp:7:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
Linking CXX executable cmTC_abca4
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_abca4.dir/link.txt --verbose=1
/opt/rh/devtoolset-11/root/usr/bin/c++ -O3 -fopenmp -funroll-loops -march=native  /usr/lib64/libopenblas.so /usr/lib/libmpicxx.so /usr/lib/libmpi.so -Wl,-rpath -Wl,/usr/lib -Wl,--enable-new-dtags CMakeFiles/cmTC_abca4.dir/petsc_test_lib.cpp.o -o cmTC_abca4  /opt/palace/lib/libpetsc.a /usr/lib64/libm.so -lstdc++ -lquadmath /usr/lib64/libdl.so /usr/lib/libmpicxx.so /usr/lib/libmpi.so
gmake[4]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
gmake[3]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'


-- PETSC_TEST_OUTPUT:
-- PETSc test program with static linkage - Success
-- Checking for module 'SLEPc'
--   No package 'SLEPc' found
-- Checking for module 'slepc'
--   Found slepc, version 3.19.99
-- Found SLEPc: 3.19.99
-- SLEPC_TEST_COMPILE_OUTPUT: Change Dir: /opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/local/bin/cmake -E env VERBOSE=1 /opt/rh/devtoolset-11/root/usr/bin/gmake -f Makefile cmTC_bc383/fast && gmake[3]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
/opt/rh/devtoolset-11/root/usr/bin/gmake  -f CMakeFiles/cmTC_bc383.dir/build.make CMakeFiles/cmTC_bc383.dir/build
gmake[4]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_bc383.dir/slepc_test_lib.cpp.o
/opt/rh/devtoolset-11/root/usr/bin/c++  -isystem /opt/palace/include -O3 -fopenmp -funroll-loops -march=native  -std=c++17 -fopenmp -o CMakeFiles/cmTC_bc383.dir/slepc_test_lib.cpp.o -c /opt/src/palace/build/palace-build/CMakeFiles/try_run/slepc_test_lib.cpp
/opt/src/palace/build/palace-build/CMakeFiles/try_run/slepc_test_lib.cpp:8:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
    8 |   PetscCall(SlepcInitialize(&argc, &argv, PETSC_NULL, PETSC_NULL));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/src/palace/build/palace-build/CMakeFiles/try_run/slepc_test_lib.cpp:8:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
Linking CXX executable cmTC_bc383
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_bc383.dir/link.txt --verbose=1
/opt/rh/devtoolset-11/root/usr/bin/c++ -O3 -fopenmp -funroll-loops -march=native  /usr/lib64/libopenblas.so /usr/lib/libmpicxx.so /usr/lib/libmpi.so -Wl,-rpath -Wl,/usr/lib -Wl,--enable-new-dtags CMakeFiles/cmTC_bc383.dir/slepc_test_lib.cpp.o -o cmTC_bc383  /opt/palace/lib/libslepc.a /opt/palace/lib/libpetsc.a /usr/lib64/libm.so -lstdc++ -lquadmath /usr/lib64/libdl.so /usr/lib/libmpicxx.so /usr/lib/libmpi.so
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: CMakeFiles/cmTC_bc383.dir/slepc_test_lib.cpp.o: in function `main':
slepc_test_lib.cpp:(.text.startup+0x24): undefined reference to `SlepcInitialize'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: slepc_test_lib.cpp:(.text.startup+0x3b): undefined reference to `EPSCreate'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: slepc_test_lib.cpp:(.text.startup+0x4d): undefined reference to `EPSDestroy'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: slepc_test_lib.cpp:(.text.startup+0x5a): undefined reference to `SlepcFinalize'
collect2: error: ld returned 1 exit status
gmake[4]: *** [CMakeFiles/cmTC_bc383.dir/build.make:105: cmTC_bc383] Error 1
gmake[4]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
gmake[3]: *** [Makefile:127: cmTC_bc383/fast] Error 2
gmake[3]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'


-- SLEPC_TEST_OUTPUT:
-- SLEPc test program - Failed
-- SLEPC_TEST_COMPILE_OUTPUT: Change Dir: /opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/local/bin/cmake -E env VERBOSE=1 /opt/rh/devtoolset-11/root/usr/bin/gmake -f Makefile cmTC_e9ca0/fast && gmake[3]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
/opt/rh/devtoolset-11/root/usr/bin/gmake  -f CMakeFiles/cmTC_e9ca0.dir/build.make CMakeFiles/cmTC_e9ca0.dir/build
gmake[4]: Entering directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_e9ca0.dir/slepc_test_lib.cpp.o
/opt/rh/devtoolset-11/root/usr/bin/c++  -isystem /opt/palace/include -O3 -fopenmp -funroll-loops -march=native  -std=c++17 -fopenmp -o CMakeFiles/cmTC_e9ca0.dir/slepc_test_lib.cpp.o -c /opt/src/palace/build/palace-build/CMakeFiles/try_run/slepc_test_lib.cpp
/opt/src/palace/build/palace-build/CMakeFiles/try_run/slepc_test_lib.cpp:8:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
    8 |   PetscCall(SlepcInitialize(&argc, &argv, PETSC_NULL, PETSC_NULL));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/src/palace/build/palace-build/CMakeFiles/try_run/slepc_test_lib.cpp:8:13: warning: PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
Linking CXX executable cmTC_e9ca0
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_e9ca0.dir/link.txt --verbose=1
/opt/rh/devtoolset-11/root/usr/bin/c++ -O3 -fopenmp -funroll-loops -march=native  /usr/lib64/libopenblas.so /usr/lib/libmpicxx.so /usr/lib/libmpi.so -Wl,-rpath -Wl,/usr/lib -Wl,--enable-new-dtags CMakeFiles/cmTC_e9ca0.dir/slepc_test_lib.cpp.o -o cmTC_e9ca0  /opt/palace/lib/libslepc.a /opt/palace/lib/libpetsc.a /usr/lib64/libm.so -lstdc++ -lquadmath /usr/lib64/libdl.so /usr/lib/libmpicxx.so /usr/lib/libmpi.so
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: CMakeFiles/cmTC_e9ca0.dir/slepc_test_lib.cpp.o: in function `main':
slepc_test_lib.cpp:(.text.startup+0x24): undefined reference to `SlepcInitialize'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: slepc_test_lib.cpp:(.text.startup+0x3b): undefined reference to `EPSCreate'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: slepc_test_lib.cpp:(.text.startup+0x4d): undefined reference to `EPSDestroy'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld: slepc_test_lib.cpp:(.text.startup+0x5a): undefined reference to `SlepcFinalize'
collect2: error: ld returned 1 exit status
gmake[4]: *** [CMakeFiles/cmTC_e9ca0.dir/build.make:105: cmTC_e9ca0] Error 1
gmake[4]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'
gmake[3]: *** [Makefile:127: cmTC_e9ca0/fast] Error 2
gmake[3]: Leaving directory '/opt/src/palace/build/palace-build/CMakeFiles/try_run/CMakeFiles/CMakeTmp'


-- SLEPC_TEST_OUTPUT:
-- SLEPc test program with static linkage - Failed
CMake Error at CMakeLists.txt:116 (message):
  SLEPc could not be found, be sure to set SLEPC_DIR
```

**Comment by sebastiangrimberg** [comment]
Hm, I am not sure why SLEPc is not being linked correctly. The link line looks fine, and we aren't getting errors that the library `/opt/palace/lib/libslepc.a` does not exist. Was there anything suspicious in the output when SLEPc was being built as part of the Palace build?

**Comment by nikosavola** [comment]
Not really, unless this gmake[7] warning here
```
==========================================
/opt/rh/devtoolset-11/root/usr/bin/gmake --print-directory -f gmakefile -l36.0 --output-sync=recurse V= slepc_libs
Now to install the library do:
make SLEPC_DIR=/opt/src/palace/build/extern/petsc/arch-linux-c-opt/externalpackages/git.slepc PETSC_DIR=/opt/palace install
=========================================
*** Installing SLEPc ***
gmake[7]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
*** Installing SLEPc at prefix location: /opt/palace  ***
====================================
Install complete.
Now to check if the libraries are working do (in current directory):
make SLEPC_DIR=/opt/palace PETSC_DIR=/opt/palace PETSC_ARCH="" check
====================================
```

**Comment by nikosavola** [comment]
I think this maybe got fixed with #56. I'll test a bit and make PR at some point.

**Comment by sebastiangrimberg** [comment]
Nice! That's what I was hoping, let me know how things go.
